glib-macros/tests/props.rs

```diff
@@ -0,0 +1,57 @@
+use glib::prelude::*;
+#[test]
+fn props() {
+    mod foo {
+        use glib::prelude::*;
+        use glib::subclass::prelude::*;
+        use glib_macros::Props;
+        use std::cell::RefCell;
+
+        mod imp {
+            use super::*;
+
+            #[derive(Props, Default)]
```

Comment:
It would be good to allow deriving this on a separate struct. That's the pattern used in all the GStreamer plugins, see e.g. [here](https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/blob/763ad0cb184e9b68bd215c9e5fe63f391a581e2d/audio/audiofx/src/audioecho/imp.rs#L39). Those are then also not using `RefCell` or the like inside the struct but only for storing the whole struct.

The reason there is that you usually want to get a snapshot of all property values at a specific time and then use that snapshot for one round of calculations, instead of possibly changing values.

Reply:
I see, that's useful

Reply:
You can do it with the latest API:

```rust
#[derive(Default)]
struct Author {
    name: String
}


#[derive(Props, Default)]
pub struct Foo {
    author: RefCell<Author>,
    #[prop(get = Self::get_author_name)]
    author_name: PhantomData<String>,
}
```

`PhantomData<T>` is needed because the macro must choose the correct `ParamSpec` for `T`.
By setting a custom getter, you can return anything you want, even another internal field

Reply:
How would that work with multiple properties for the same struct? You would just add multiple phantom data and in the getter/setter function you do the right thing then?

Reply:
Maybe there could be some additional convenience though. You could allow something like `self.author.name` as getter/setter and the macro does the refcell + field access itself?

Maybe something like

```rust
#[derive(Default)]
struct Author {
    first_name: String,
    last_name: String,
}


#[derive(Props, Default)]
pub struct Foo {
    author: RefCell<Author>,
    #[prop(delegate = self.author.first_name)]
    author_first_name: String,
    #[prop(delegate = self.author.last_name)]
    author_last_name: String,
}
```

and whenever you see a `delegate` you would remove the field from the struct after macro expansion. That way you wouldn't need the `PhantomData`. Not sure if you can remove fields from derive macros. The alternative would be to make them comments (you can parse comments from proc macros), e.g.

```rust
#[derive(Props, Default)]
pub struct Foo {
    author: RefCell<Author>,
    #[prop(delegate = self.author.first_name)]
    // author_first_name: String,
    #[prop(delegate = self.author.last_name)]
    // author_last_name: String,
}
```

Reply:
I don't know... I feel that's a layer of abstraction too high...
To implement `delegate` I need to create a `Monad`-like abstraction over all the types implementing interior mutability, because I need to support `RefCell`, `Mutex`, `RwLock`, [...].

I feel this goes out of the scope "create a glib property easily".

With the current API you can already do this, inline:
`get = |t: &Self| t.author.borrow().first_name.to_value()`

It's long, but I feel that's not gtk-rs fault.

I don't want to add too much "magic" behind the scenes...

Anyway, I'll keep your comments in mind while I try to implement the missing things

Reply:
> To implement delegate I need to create a Monad-like abstraction over all the types implementing interior mutability, because I need to support RefCell, Mutex, RwLock, [...].

Well, you would need a trait like

```rust
trait DelegateStorage {
    type Inner;

    fn set(&self, func: impl FnOnce(&mut T));
    fn get(&self, func: impl FnOnce(&T));
}

impl<T> DelegateStorage for Rc<T> {
    type Inner = T;
    [...]
}
```

Doesn't seem too hard but reduces quite a bit of boilerplate :) But I agree that this could be added in a second step.

Reply:
> But I agree that this could be added in a second step

... but ideally before merging this. Because otherwise updating gst-plugins-rs to the new API would result in more code than with the old one.

Reply:
What do you think of this API?
```rust
#[prop(name = "author-name", get, set, type = String, member = name)]
#[prop(name = "author-nick", get, set, type = String, member = nick)]
author: RefCell<Author>,
```

It doesn't require creating a fake field. I think it's more practical and feels nicer. The two props share the same underlying store, a single `RefCell`, so I like them being defined on the same field

Reply:
That's nice, I like it :) The only sad part is that the type has to be duplicated but I guess that can't really be avoided